### ISSUE TITLE
Using switch_logits of final doc_span only

### ISSUE THREAD

**Issue by talkhaldi** [AkariAsai/learning_to_retrieve_reasoning_paths#5]
Hi again,

I'm reading through rc_utils.py and I found that you only use the switch_logits of the final doc_span/stride. Is this intentional?

[https://github.com/AkariAsai/learning_to_retrieve_reasoning_paths/blob/master/reader/rc_utils.py#L1019](url)

You also use it in lines 1061 and 1062. Have you tried other ways? Like taking average of switch logits over all doc_spans, or max.. etc.?

**Comment by AkariAsai** [comment]
Hi, thank you for your interest! 

> https://github.com/AkariAsai/learning_to_retrieve_reasoning_paths/blob/master/reader/rc_utils.py#L1019
You also use it in lines 1061 and 1062.

The [rc_utils.py#L1019](https://github.com/AkariAsai/learning_to_retrieve_reasoning_paths/blob/master/reader/rc_utils.py#L1019) picks up `no_answer` probability. Our reader model is a multi-task learner of a span extraction task and a 4-way classification task of (0) `span-answer`, (1) `no_answer`, (2) `yes` and (3) `no`. We re-rank the reasoning paths based on `no_answer` probability. 

In addition, we keep `switch`, and if the `switch==2` or `switch==3`, we output yes or no instead of the spans, which is handled in [switch_answer](https://github.com/AkariAsai/learning_to_retrieve_reasoning_paths/blob/4c78776065409f6faaa15337bde44a9fccc3f887/reader/rc_utils.py#L1348) function.

> you only use the switch_logits of the final doc_span/stride. 

As far as I remember, the `unique_id` corresponds to each stride and the `result`, defined [here](https://github.com/AkariAsai/learning_to_retrieve_reasoning_paths/blob/4c78776065409f6faaa15337bde44a9fccc3f887/reader/rc_utils.py#L964), stores the span results  & classification result for each stride. 
So`no_answer_logit = result.switch_logits[1]` will take the `no_answer` probability for the stride.  We then [aggregate and re-rank the spans from all strides](https://github.com/AkariAsai/learning_to_retrieve_reasoning_paths/blob/4c78776065409f6faaa15337bde44a9fccc3f887/reader/rc_utils.py#L1012)  based on their span probabilities following the huggingface transformer's [span selection method at that time](https://github.com/huggingface/transformers/blob/v0.6.2/examples/run_squad.py#L441). In other words, the `no_answer` probability ( `switch_logits[1]` ) is from the stride that contains the most possible answer span in the reasoning path.  

> Have you tried other ways? Like taking average of switch logits over all doc_spans, or max.. etc.?

I think taking the average of switch logits from all of the strides can be noisy in some cases. For example, there might be some cases like: we have a long reasoning path that can be divided into several strides, and one or some of the strides are not quite relevant to the question. Our reasoning path reader can output high `no_answer` probabilities for these irrelevant strides. Even if one of the strides is quite relevant to the question and has the correct answer, the reasoning path is likely to be ranked lower, as we take the average of the `no_answer` probabilities across the strides.

We also would like to note that the majority of the cases, our input (reasoning path) for each question is within a single window.

**Comment by talkhaldi** [comment]
Thank you for your explanation!

I understand, but I want to point out one thing about this:

> Sono_answer_logit = result.switch_logits[1] will take the no_answer probability for the stride.

This is correct, but looking at the indentation of line 1019, I see that the "result" you're using is the one that is left over from the last iteration of the loop at line 963. If you want to consider the `no_answer_logit` of the highest ranked stride (based on span logits), I suggest the following at line 1019:

`  feature_of_highest_stride = features[prelim_predictions[0].feature_index] `
` result_of_highest_stride = unique_id_to_result[feature_of_highest_stride] `
` no_answer_logit = result_of_highest_stride.switch_logits[1]`

Is my understanding correct?

Also, this won't matter much as the majority of the cases have one stride as you say.

**Comment by AkariAsai** [comment]
Oh, thanks for catching that! You're right. Sorry for my misunderstanding of your previous post. 

We'll take a closer look at that issue, and may update our codebase - before changing our inference codes, we would like to re-evaluate our code on all datasets to make sure it would not change the results much.

**Comment by talkhaldi** [comment]
Got it! No worries.
Good luck!
